### PR TITLE
Reducing the re-rendering of the chat input

### LINKED ISSUE #1679: Blob from text editor shows
Not sure when this happens, but I have experienced it a couple of times.

https://github.com/acterglobal/a3/assets/77483554/b72454bf-ca72-4c5c-89c8-c6a7a1acae4b

### PR DESCRIPTION
**Pull request by gnunicorn**
We were still re-rendering that way to often. This refactors this so that we rarely re-render the chat input box while we are in the chat screen. We only rerender it now for when we are a) loading in the first place or b) the permissions have changed and disallow us to send messages now. Additionally this improves the loading animation of chat rooms in the list and the room itself by some smarter using of our providers:

https://github.com/acterglobal/a3/assets/40496/19341588-7b48-45c2-a8da-9147d52d3283

Fixes #1679 , and probably also fixes https://github.com/acterglobal/a3-meta/issues/246 .

### CODE REVIEW

File: app/lib/common/providers/room_providers.dart [hunk 1]
```diff
@@ -279,8 +279,7 @@ final relatedSpacesProvider = FutureProvider.autoDispose
 
 /// Get the user's membership for a specific space based off the spaceId
 /// will throw if the client doesn't kow the space
-final roomMembershipProvider =
-    FutureProvider.autoDispose.family<Member?, String>(
+final roomMembershipProvider = FutureProvider.family<Member?, String>(
```
Comment (gnunicorn) [inline app/lib/common/providers/room_providers.dart:282]: avoids rerenders

File: app/lib/features/chat/pages/room_page.dart [hunk 2]
```diff
@@ -42,15 +42,107 @@ class RoomPage extends ConsumerWidget {
     super.key = roomPageKey,
   });
 
+  Widget appBar(BuildContext context, WidgetRef ref) {
```
Comment (gnunicorn) [inline app/lib/features/chat/pages/room_page.dart:45]: just a move of the function

File: app/lib/features/chat/pages/room_page.dart [hunk 3]
```diff
@@ -42,15 +42,107 @@ class RoomPage extends ConsumerWidget {
...
+                    child: Text(L10n.of(context).loadingRoomFailed(e)),
+                  ),
+                  loading: () => Center(child: Text(L10n.of(context).loading)),
+                ),
+            chatInput(context, ref),
+          ],
+        ),
+      ),
+    );
+  }
```
Comment (gnunicorn) [inline app/lib/features/chat/pages/room_page.dart:128]: rendering appbar and chatinput outside of the chatProvider watching as they don't actually need it, they are happy with just the roomId

File: app/lib/features/chat/pages/room_page.dart [hunk 4]
```diff
@@ -42,15 +42,107 @@ class RoomPage extends ConsumerWidget {
...
+
+  Widget chatInput(BuildContext context, WidgetRef ref) {
```
Comment (gnunicorn) [inline app/lib/features/chat/pages/room_page.dart:130]: another move

File: app/lib/features/chat/pages/room_page.dart [hunk 5]
```diff
@@ -42,15 +42,107 @@ class RoomPage extends ConsumerWidget {
...
+      userAppSettingsProvider.select(
+        (settings) => settings.valueOrNull?.typingNotice() ?? false,
+      ),
+    );
+    return CustomChatInput(
+      key: Key('chat-input-$roomId'),
+      roomId: roomId,
+      onTyping: sendTypingNotice
+          ? (typing) async {
+              (await ref.read(chatProvider(roomId).future))
```
Comment (gnunicorn) [inline app/lib/features/chat/pages/room_page.dart:141]: only this line actually changed.

File: app/lib/features/chat/pages/room_page.dart [hunk 6]
```diff
@@ -114,152 +205,121 @@ class _ChatRoomConsumerState extends ConsumerState<ChatRoom> {
...
-        Widget child, {
-        required types.Message message,
-        required bool nextMessageInGroup,
-      }) =>
-          GestureDetector(
-        onSecondaryTap: () {
-          showMessageOptions(context, message, roomId);
-        },
-        child: BubbleBuilder(
+    return Expanded(
```
Comment (gnunicorn) [inline app/lib/features/chat/pages/room_page.dart:214]: only wrapped in the `Expanded(`

File: app/lib/features/chat/widgets/custom_input.dart [hunk 7]
```diff
@@ -58,34 +58,51 @@ class CustomChatInput extends ConsumerWidget {
...
+            membership.valueOrNull?.canString('CanSendChatMessages'),
+      ),
+    );
+    if (canSend == null) {
+      // we are still loading
+      return loadingState(context);
+    }
+    if (canSend) {
+      return _ChatInput(roomId: roomId, onTyping: onTyping);
+    }
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/custom_input.dart:73]: reducing rerenders via `.select`

File: app/lib/common/widgets/chat/convo_with_avatar_card.dart [hunk 8]
```diff
@@ -1,12 +1,16 @@
...
 import 'package:acter/common/themes/colors/color_scheme.dart';
 import 'package:acter/features/chat/providers/chat_providers.dart';
 import 'package:acter/features/chat/widgets/room_avatar.dart';
 import 'package:acter_avatar/acter_avatar.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:skeletonizer/skeletonizer.dart';
 
 class ConvoWithAvatarInfoCard extends ConsumerWidget {
+  final String? title;
```
Comment (gtalha07) [inline app/lib/common/widgets/chat/convo_with_avatar_card.dart:13]: where are using this down in a widget??? seems not mentioned anywhere...

File: app/lib/common/widgets/chat/convo_with_avatar_card.dart [hunk 9]
```diff
@@ -41,10 +45,12 @@ class ConvoWithAvatarInfoCard extends ConsumerWidget {
     this.subtitle,
     this.trailing,
     this.showParents = true,
+    this.title,
```
Comment (gnunicorn) [inline app/lib/common/widgets/chat/convo_with_avatar_card.dart:48]:
```suggestion
```

File: app/lib/common/widgets/chat/convo_with_avatar_card.dart (hunk 8 above)
Comment (gnunicorn) [inline app/lib/common/widgets/chat/convo_with_avatar_card.dart:13]:
```suggestion
```

Comment (gnunicorn) [reply app/lib/common/widgets/chat/convo_with_avatar_card.dart:13]: yeah, no that was an in-between attempt that got abandoned ...

**Comment by emilvincentz** [comment]
@gnunicorn just chekcing if has this been merged into the latest app release?
The bug still appears

**Pull request merged**